File: packages/9router/src/providers/groq.ts
import Groq from 'groq-sdk';
import { BaseProvider } from './base.js';
import { Message } from '../types.js';

export class GroqProvider extends BaseProvider {
  name = 'groq';
  private client: Groq;

  constructor(apiKey: string) {
    super();
    this.apiKey = apiKey;
    this.client = new Groq({ apiKey });
  }

  async call(
    messages: Message[],
    modelId: string,
    options?: {
      maxTokens?: number;
      temperature?: number;
      timeout?: number;
    },
  ): Promise<{
    text: string;
    inputTokens: number;
    outputTokens: number;
  }> {
    try {
      // Groq is OpenAI-compatible: the system prompt is a system-role message,
      // not a top-level param. Pass all messages through in order.
      const response = await this.client.chat.completions.create({
        model: modelId,
        max_tokens: options?.maxTokens || 4096,
        temperature: options?.temperature || 0.7,
        messages: messages.map((m) => ({
          role: m.role,
          content: m.content,
        })) as any,
      });

      const textContent = response.choices[0]?.message.content;
      if (!textContent) {
        throw new Error('No text content in response');
      }

      return {
        text: textContent,
        inputTokens: response.usage?.prompt_tokens || 0,
        outputTokens: response.usage?.completion_tokens || 0,
      };
    } catch (error) {
      throw this.parseError(error);
    }
  }
}

File: betterprompts/betterprompts/providers/groq.py
from groq import Groq
from .base import ProviderError, RateLimitError


class GroqProvider:
    def __init__(self, api_key: str, model: str = "llama-3.3-70b-versatile"):
        if not api_key:
            raise ProviderError("Groq API key not provided")
        self.api_key = api_key
        self.model = model
        self.client = Groq(api_key=api_key)

    def improve(self, system: str, user: str) -> str:
        try:
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
            )
            if response.choices and response.choices[0].message.content:
                return response.choices[0].message.content
            raise ProviderError("Groq returned empty response")
        except Exception as e:
            error_str = str(e).lower()
            if "rate_limit" in error_str or "quota" in error_str:
                raise RateLimitError(f"Groq rate limit exceeded: {e}")
            raise ProviderError(f"Groq error: {e}")
